Plotting recipes for geo arrays with XYZ dimensions

ali-ramadhan opened the pull request:
This might be a controversial PR since it modifies the plotting recipes quite a bit, especially with a focus on plotting geo arrays with XYZ dimensions.

Happy to work on this PR to get something that works nicely for both Lat/Lon and XYZ plots.

This PR also introduces `GeoXDim`, `GeoYDim`, and `GeoZDim` for `Lat`/`Lon` to subtype following @rafaqz's suggestion in #103.

Resolves #103

Some examples to showcase the motivation behind my changes (and this PR):

# xy plot

```
GeoArray with dimensions:
 x (type xC): Float64[20.0, 60.0, …, 2500.0, 2540.0] (Sampled: Ordered Regular Intervals) (length 64) 
 y (type yC): Float64[20.0, 60.0, …, 2500.0, 2540.0] (Sampled: Ordered Regular Intervals) (length 64) 
and referenced dimensions:
 Time (type Ti): 7330.000000067156 (Sampled: Ordered Irregular Intervals)
 z (type zC): 1260.0 (Sampled: Ordered Regular Intervals)
```

![image](https://user-images.githubusercontent.com/20099589/100175719-ecf22480-2e9c-11eb-8ce7-b1592a5ea8f8.png)

# yz plot

```
GeoArray with dimensions:
 y (type yC): Float64[20.0, 60.0, …, 2500.0, 2540.0] (Sampled: Ordered Regular Intervals) (length 64) 
 z (type zC): Float64[20.0, 60.0, …, 2500.0, 2540.0] (Sampled: Ordered Regular Intervals) (length 64) 
and referenced dimensions:
 Time (type Ti): 7330.000000067156 (Sampled: Ordered Irregular Intervals)
 x (type xC): 620.0 (Sampled: Ordered Regular Intervals)

```

![image](https://user-images.githubusercontent.com/20099589/100175816-1e6af000-2e9d-11eb-9fad-d920de6a32f5.png)

# xz plot

```
GeoArray with dimensions:
 x (type xC): Float64[20.0, 60.0, …, 2500.0, 2540.0] (Sampled: Ordered Regular Intervals) (length 64) 
 z (type zC): Float64[20.0, 60.0, …, 2500.0, 2540.0] (Sampled: Ordered Regular Intervals) (length 64) 
and referenced dimensions:
 Time (type Ti): 7330.000000067156 (Sampled: Ordered Irregular Intervals)
 y (type yC): 2540.0 (Sampled: Ordered Regular Intervals)
```

![image](https://user-images.githubusercontent.com/20099589/100175912-49554400-2e9d-11eb-8a19-41e905f9c721.png)

# Vertical profiles

```
GeoArray with dimensions:
 z (type zC): Float64[20.0, 60.0, …, 2500.0, 2540.0] (Sampled: Ordered Regular Intervals) (length 64) 
and referenced dimensions:
 Time (type Ti): 7330.000000067156 (Sampled: Ordered Irregular Intervals)
 x (type xC): 620.0 (Sampled: Ordered Regular Intervals)
 y (type yC): 1780.0 (Sampled: Ordered Regular Intervals)
```

![image](https://user-images.githubusercontent.com/20099589/100176073-a18c4600-2e9d-11eb-9beb-7111ef524816.png)

# Stacked plots

```
GeoArray with dimensions:
 x (type xC): Float64[20.0, 60.0, …, 2500.0, 2540.0] (Sampled: Ordered Regular Intervals) (length 64) 
 y (type yC): Float64[20.0, 60.0, …, 1220.0, 1260.0] (Sampled: Ordered Regular Intervals) (length 32) 
 z (type zC): Float64[20.0, 340.0, …, 1940.0, 2260.0] (Sampled: Ordered Regular Intervals) (length 8) 
and referenced dimensions:
 Time (type Ti): 7330.000000067156 (Sampled: Ordered Irregular Intervals)
```

![image](https://user-images.githubusercontent.com/20099589/100176250-f203a380-2e9d-11eb-9be1-2ff065779394.png)

ali-ramadhan commented:
Note to self: Should use units from NetCDF metadata to label axes and colorbars.

rafaqz commented on src/plotrecipes.jl:
```diff
@@ -1,60 +1,64 @@
-@recipe function f(::GeoPlot, A::GeoArray{T,2,<:Tuple{<:Lon,<:Lat}}) where T
-    GeoPlot(), permutedims(A)
+# # Plot a vertical 1d line
+@recipe function f(::GeoPlot, A::GeoArray{T,1,<:Tuple{<:GeoZDim}}) where T
```
Does this need to be in GeoData? should all DimensionalData `Z` plot vertically? (I have no idea)

ali-ramadhan replied:
Ah I'm not sure but at least in oceanography and atmospheric science it's very common to want to plot a vertical profile (vector or 1D array) of some quantity, in which case it's nice to have the vertical coordinate be on the y-axis so the profile is oriented in an intuitive way (see the vertical profile plot I posted above).

By default the z-coordinate will go from left to right which makes interpreting vertical profiles a little tougher (unless you tilt your head sideways 🙃).

rafaqz commented on src/plotrecipes.jl:
```diff
@@ -1,60 +1,64 @@
+    :xguide --> name(dims(A, 1))
+    :yguide --> name(dims(A, 2))
+    x1, x2 = map(prepare, dims(A))
+    x1, x2, permutedims(A) |> parent
```
Now that you generalised this method, we will need to specify the exact permutation we want to have plotted:

```julia
permutedims(A, (GeoXDim, GeoYDim)) 
```
(it might be the other way around I don't remember)

You will also need to handle `Z` dim. If for example `X` is always first you can probably do:

```julia
permutedims(A, (GeoXDim, Dimension)) 
```

Which will work for both.

rafaqz commented:
This looks great. My main concern is to make sure we keep tiling of the time dimension- so that a `Lat, Lon, Ti` is always tiled heatmaps - one for each slice in time. 

So for clarity, is it fair to say that the rules would be:
- GeoXDim, GeoYDim: Heatmap
- GeoXDim, GeoYDim, Ti: Tiles in Ti o X/Y heatmap with X on X dimension of plot
- GeoXDim, GeoYDim, GeoZDim:  Tiles in Z of X/Y heatmap with X on X dim of plot ???
- GeoYDim, GeoZDim: Heatmap, Y on Y dimension of plot, Z on X?
- GeoXDim, GeoZDim: Heatmap, X on X dimension of plot, Z on Y?
- GeoXDim: Line plot with X on the X dimension of plot?
- GeoZDim: Line plot with Y on the ??? dimension of plot?
- GeoZDim: Line plot with Z on the Y dimension of plot?

And the original permutation of X, Y, Z, Ti should make no difference to the final plot.

rafaqz commented:
Also I'm using GeoData.jl for making plots in a paper. Unfortunately the defaults don't look great... 

I'm thinking we canm use the style I work out for paper in the recipes, and start trying to make everything look clean and publication worthy out of the box.

rafaqz commented on src/plotrecipes.jl:
```diff
@@ -1,60 +1,64 @@
-                lats, lons = map(prepare, dims(slice))
-                lons, lats, parent(slice)
+                xs, ys = map(prepare, dims(slice))
+                xs, ys, permutedims(slice) |> parent
```
This will also need the same permutedims method. Can probably use the same method for both.

rafaqz commented on src/plotrecipes.jl:
```diff
@@ -1,60 +1,64 @@
 
 # Plot 3d arrays as multiple tiled plots
-@recipe function f(::GeoPlot, A::GeoArray{T,3,<:Tuple{<:Lat,<:Lon,D}}) where {T,D}
+@recipe function f(::GeoPlot, A::GeoArray{T,3,<:Tuple{<:GeoDim,<:GeoDim,D}}) where {T,D}
```
This wont handle cases correctly where e.g the GeoXDim is the third dimension.

ali-ramadhan replied:
Ah hmmm is this because we cannot assume the dimensions are in any order, e.g. X, Y, Z, T? How would you suggest I modify the function signature to handle all cases correctly?

rafaqz replied:
Yeah. Before I assumed Lat/Lon were the first two dims, but could be in either order. So the map always plots the right way.

Adding Z as well we probably can't assume anything now... so we can instead eagerly permute to the order that we want and only define one function signature with GeoPlot - the order will be fixed.

rafaqz commented:
Ok heres a simple one-liner to sort out all of the permutations right at the start - it can probably be in the `prepare` method in the initial recipe.

```julia
 permutedims(da, DimensionalData.commondims((GeoXDim, GeoYDim, GeoZDim, TimeDim), dims(da)))
```

There is a little bug in `commondims` that stops us from cleanly using `Dimension` as a catchall at the end instead of `TimeDim`.  Will fix.

ali-ramadhan commented:
Thank you for the feedback @rafaqz! And sorry it took a while to get back to this PR! Added some more niceties and addressed most of your comments I think.

> Also I'm using GeoData.jl for making plots in a paper. Unfortunately the defaults don't look great...
> 
> I'm thinking we can use the style I work out for paper in the recipes, and start trying to make everything look clean and publication worthy out of the box.

That would be awesome! I'm happy to help with this iterative process if I can.

I think my priority is to simplify plotting and animation for Oceananigans.jl simulation output by switching to GeoData.jl. There's currently quite a bit of boilerplate, e.g. https://clima.github.io/OceananigansDocumentation/stable/generated/two_dimensional_turbulence/#Visualizing-the-results and https://github.com/CliMA/LESbrary.jl/blob/b22e405cb03ac9f25dd0053d1ab9420b3ac8fb2e/examples/three_layer_constant_fluxes.jl#L463-L581

> So for clarity, is it fair to say that the rules would be:
> 
>   * GeoXDim, GeoYDim, GeoZDim, Dimension: Not defined

I think this makes sense.

>   * GeoXDim, GeoYDim, Dimension: Tiles in Dimension of X/Y heatmap with X on X dimension of plot
>   * GeoXDim, GeoYDim, GeoZDim: Tiles in Z of X/Y heatmap with X on X dim of plot ???

I think these make sense but aren't they the same case?

>   * GeoXDim, GeoYDim: Heatmap, X on X dimension of plot, Y on Y
>   * GeoYDim, GeoZDim: Heatmap, Y on X dimension of plot, Z on Y?
>   * GeoXDim, GeoZDim: Heatmap, X on X dimension of plot, Z on Y?

Yes I guess the idea is to put the Z dimension on the y-axis of the plot if possible, otherwise put y.

>   * GeoXDim: Line plot with X on the X dimension of plot?
>   * GeoYDim: Line plot with Y on the ??? dimension of plot?
>   * GeoZDim: Line plot with Z on the Y dimension of plot?

Yes I think for `GeoZDim` it should be a line plot with Z on the Y dimension of the plot, but for `GeoXDim` and `GeoYDim` it should be on the X dimension of the plot.

This is certainly field-dependent but I guess coming from oceanography/atmospheric science X is sometimes the zonal direction while Y is the meridional direction (both are always horizontal dimensions) so they get plotted on the x-axis of the plot. But Z is a vertical dimension so it gets plotted on the y-axis of the plot.

ali-ramadhan commented:
Axis and colorbar labels will now include the units if they're defined in the metadata.

I also added a feature that uses a symmetric colorbar `:balance` if it detects that the data plotted is centered around zero and the extreme ends differ by less than 25%. This is useful for plotting things like velocities.

I also added some logic for prettying up `contourf` plots!

![image](https://user-images.githubusercontent.com/20099589/101272454-3e7e9700-375a-11eb-8014-af1f9d93ab37.png)

Not sure if all the changes I made are appropriate, happy to iterate though!

ali-ramadhan commented:
Hmmm looks like I get warnings like this when plotting (not sure if I did something wrong)

```
[2020/12/06 18:29:31.225] WARN  (xF, yC, zC) dims were not found in object -@-> /home/alir/.julia/packages/DimensionalData/fTd6d/src/primitives.jl:565
```

rafaqz commented:
This is looking good. I'm really busy with DynamicGrids things right now, so might be a bit slow to respond. 

I also recognise that getting these permutations right and totally generalised is hard enough that I didn't bother to do it properly the first time, and might be a big ask. I know how to do it now so can get this finished if that is easier for you - it might just take a while for me to have time.

ali-ramadhan commented:
That's quite all right! No rush on my part. I'm just depending on this branch as I test out the new plotting recipes so it's easier to make adjustments.

Yeah not sure if the permutations are completely correct right now but I might play around with it to get some plots to look right.

Some eye candy made with GeoData.jl plotting recipes!

![image](https://user-images.githubusercontent.com/20099589/101360585-d61fda80-386b-11eb-8811-a011b96c4f04.png)

rafaqz commented:
Those plots look amazing

rafaqz commented:
FYI this is just waiting on the next release of DD to be possible in a fully generic way.
